be/src/olap/tablet.cpp

```diff
@@ -358,9 +383,102 @@ void Tablet::delete_expired_inc_rowsets() {
+
+        auto first_version = to_delete_version.front();
+        auto end_version = to_delete_version.back();
+        Version test_version = Version(first_version->version().first, end_version->version().second);
+
+        OLAPStatus status = capture_consistent_versions(test_version, nullptr);
+        // When there is no consistent versions, we must reconstruct the tracker.
+        if (status != OLAP_SUCCESS) {
+            LOG(WARNING) << "The consistent version check fails, there are bugs. "
+                << "Reconstruct the tracker to recover versions in tablet=" << tablet_id();
+
+            _timestamped_version_tracker.reconstruct_versioned_tracker(
+                    _tablet_meta->all_rs_metas(), expired_snapshot_version_path_map);
+
+            // double check the consistent versions
+            status = capture_consistent_versions(test_version, nullptr);
+
+            if (status != OLAP_SUCCESS) {
+                if (!config::ignore_rowset_expired_snapshot_unconsistent_delete) {
+                    LOG(FATAL) << "rowset expired snapshot unconsistent delete. tablet= " << tablet_id();
+                } else {
+                    LOG(WARNING) << "rowset expired snapshot unconsistent delete. tablet= " << tablet_id();
+                }
+            }
+            return;
+        }
+        expired_snapshot_version_path_map[*path_id_iter] = version_path;
+        path_id_iter++;
+    }
+
+    auto old_size = _expired_snapshot_rs_version_map.size();
+    auto old_meta_size = _tablet_meta->all_expired_snapshot_rs_metas().size();
+
+    // do delete operation
+    auto to_delete_iter = expired_snapshot_version_path_map.begin();
+    while (to_delete_iter != expired_snapshot_version_path_map.end()) {
+        
+        std::vector<TimestampedVersionSharedPtr>& to_delete_version = to_delete_iter->second->timestamped_versions();
+        for (auto& timestampedVersion : to_delete_version) {
+            auto it = _expired_snapshot_rs_version_map.find(timestampedVersion->version());
+            if (it != _expired_snapshot_rs_version_map.end()) {
+                // delete rowset
+                StorageEngine::instance()->add_unused_rowset(it->second);
+                _expired_snapshot_rs_version_map.erase(it);
+                LOG(INFO) << "delete expired snapshot rowset tablet=" << full_name() 
+                    <<" version[" << timestampedVersion->version().first << "," << timestampedVersion->version().second 
+                    << "] move to unused_rowset success " << std::fixed << expired_snapshot_sweep_endtime;
+            } else {
+                LOG(WARNING) << "delete expired snapshot rowset tablet=" << full_name() 
+                    <<" version[" << timestampedVersion->version().first << "," << timestampedVersion->version().second 
+                    << "] not find in expired snapshot rs version ";
+            }
+            _delete_expired_snapshot_rowset_by_version(timestampedVersion->version());
```

Comment:
in `_delete_expired_snapshot_rowset_by_version`, we call `_expired_snapshot_rs_version_map.erase()` again, which is duplicated.